cpp/Contact.cpp

```diff
@@ -549,6 +549,202 @@ void dolfinx_contact::Contact::assemble_vector(
   }
 }
 //-----------------------------------------------------------------------------------------------
+std::pair<std::vector<PetscScalar>, int>
+dolfinx_contact::Contact::pack_grad_test_functions(
+    int pair, const xtl::span<const PetscScalar>& gap,
+    const xtl::span<const PetscScalar>& u_packed)
+{
+  const int puppet_mt = _contact_pairs[pair][0];
+  const int candidate_mt = _contact_pairs[pair][1];
+  // Mesh info
+  const dolfinx_contact::SubMesh& submesh = _submeshes[candidate_mt];
+  std::shared_ptr<const dolfinx::mesh::Mesh> mesh = _V->mesh(); // mesh
+  assert(mesh);
+  const std::size_t gdim = mesh->geometry().dim(); // geometrical dimension
+  std::vector<std::int32_t> parent_cells = submesh.parent_cells();
+  const std::size_t ndofs = _V->dofmap()->cell_dofs(0).size();
```

Comment:
This is not safe in parallel, and should be as: https://github.com/FEniCS/dolfinx/blob/cebaf72be7642512794b0db2932617560dac57fc/cpp/dolfinx/fem/Function.h#L352